new SQL Exec Plan obfuscation API for checks

djova opened the pull request:
### What does this PR do?

Adds a new python API `datadog_agent.obfuscate_sql_exec_plan` to enable checks to obfuscate & normalize JSON Query Execution plans.

* `obfuscate_sql_exec_plan(plan)` - obfuscates the provided plan, keeping only those fields which we know are safe and can't contain any raw query information 
* `obfuscate_sql_exec_plan(plan, normalize=True)` - in addition to obfuscating away everything that regular plan obfuscation does, this also obfuscates away cost & row estimate related info to produce a normalized execution plan. 

Adds a new `jsonObfuscator` value transformation feature:
* enable transforming all values for a given set of keys using the given transformer function (only out-of-the box function now supported is `obfuscate_sql`). Used by plan obfuscation to obfuscate query conditions like `(user.id = 5)`.
* the `obfuscate_sql` function is now supported for the mongo and ES JSON obfuscators as well so if someone does have SQL they want to obfuscate in those JSON objects then they'll be able to do so

Other minor changes:
* `jsonObfuscator` tests improved to do deeper comparison of expected json and show more detailed diffs
* lazy initialize obfuscator on first use in `datadog_agent.go` so it can use the agent config. If we initialize on module init then we can't guarantee that the agent config has been loaded.

### Motivation

Follow-up to https://github.com/DataDog/datadog-agent/pull/5523. We want to enable python checks to obfuscate SQL Execution Plans in addition to raw SQL queries. 

### Describe your test plan

* unit tests for JSON obfuscator changes
* unit tests for new python API
* running successfully in staging for several days already

justiniso commented on pkg/trace/obfuscate/sql.go:
```diff
@@ -342,3 +344,73 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
+	return obj
+}
+
+var planObfuscationPattern = regexp.MustCompile(`attached_condition|(\w+ Cond)`)
```
Regex feels like the wrong solution here and prone to unexpected behavior over time, especially as we add more dbms. Is there any reason not to enumerate the list of keys in a map? I think that would make the code easier to read/modify and less prone to bugs.

- `attached_condition`
- `Index Cond`
- `Recheck Cond`
- ...

djova replied:
Good idea! Updated to use an explicit list of keys for both obfuscation and normalization.

justiniso commented on pkg/trace/obfuscate/sql.go:
```diff
@@ -342,3 +344,73 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
+	return resultMap
+}
+
+var planNormalizationPattern = regexp.MustCompile(`(\w+_cost)|rows_examined_per_scan|rows_produced_per_join|data_read_per_join|filtered|(\w Cost)|Plan Rows|Plan Width`)
```
Same comment as above on regex.

justiniso commented on pkg/trace/obfuscate/sql_plan_test.go:
```diff
@@ -0,0 +1,89 @@
+
+type execPlanTestCase struct {
+	// Rdbms is mysql or postgres
+	Rdbms          string                 `json:"rdbms"`
```
nit: Dbms is more correct. Mongodb has a query planner and is non relational.

justiniso replied:
Also nit: isn't the convention in go that all acronyms are uppercase?

djova replied:
good point, updated

justiniso commented on pkg/trace/obfuscate/sql.go:
```diff
@@ -342,3 +344,73 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
+		}
+		obfQuery, err := o.ObfuscateSQLString(strValue)
+		if err != nil {
+			log.Errorf("failed to obfuscate sql plan: %s", err.Error())
```
I think these logs could be noisy and the returned values result in false positives of a plan "changing" even when it was just an error. What do you think about instead returning an error for the whole plan for these cases? Then we would throw it away rather than submit a plan with overly-obfuscated values.

djova replied:
> I think these logs could be noisy

Agreed, updated to send debug logs instead. 

> Then we would throw it away rather than submit a plan with overly-obfuscated values.

Why is it better to throw it away than to overly obfuscate? Obfuscation failures really shouldn't happen because that would mean we're probably failing to obfuscate the full query samples that contain that condition as well. 

What do you think of putting an error message in the obfuscated value, that way the plan still gets ingested enabling us to detect that the failed obfuscation is happening? 

See here https://github.com/DataDog/datadog-agent/pull/5692/commits/40a54b3cb2bcee2b5d1bfb96366ec70c9eec4ef0

justiniso replied:
> Obfuscation failures really shouldn't happen because that would mean we're probably failing to obfuscate the full query samples that contain that condition as well.

Agreed, it would mean we missed a case, but it can and will happen once we expose it to users' queries. We need a way to discover these since this will be behind customer infrastructure. Maybe we should emit a metric.

> What do you think of putting an error message in the obfuscated value, that way the plan still gets ingested enabling us to detect that the failed obfuscation is happening?

Since that plan will go to users, I don't think we should do that. Ultimately it's an error on our part, not something they can fix. So the workflow would be for them to see a plan with that error message, enable debug logs on the agent, and then tell us what the error message says? I'm not a fan of that solution.

If we choose to show the plan (sending '?' on the key that error'ed), we need a way to signal that it's semi-correct. I would rather see the errors surfaced up to the outer function and attach it as another field to the log message so we could show a ⚠️  or something in the UI. Then it's best-effort to display and we can track error frequency.

justiniso commented on pkg/trace/obfuscate/sql_plan_test.go:
```diff
@@ -0,0 +1,89 @@
+			assert.Equal(t, testCase.NormalizedPlan, normalized)
+		})
+	}
+}
```
nit: to add benchmarking tests

djova replied:
good idea, added

djova committed:
new SQL Exec Plan obfuscation API for checks

Adds a new python API `datadog_agent.obfuscate_sql_exec_plan` to enable checks to obfuscate & normalize JSON Query Execution plans.

* `obfuscate_sql_exec_plan(plan)` - obfuscates the provided plan, keeping only those fields which we know are safe and can't contain any raw query information
* `obfuscate_sql_exec_plan(plan, normalize=True)` - in addition to obfuscating away everything that regular plan obfuscation does, this also obfuscates away cost & row estimate related info to produce a normalized execution plan.

Adds a new `jsonObfuscator` value transformation feature:
* enable transforming all values for a given set of keys using the given transformer function (only out-of-the box function now supported is `obfuscate_sql`). Used by plan obfuscation to obfuscate query conditions like `(user.id = 5)`.
* the `obfuscate_sql` function is now supported for the mongo and ES JSON obfuscators as well so if someone does have SQL they want to obfuscate in those JSON objects then they'll be able to do so

Other minor changes:
* `jsonObfuscator` tests improved to do deeper comparison of expected json and show more detailed diffs
* lazy initialize obfuscator on first use in `datadog_agent.go` so it can use the agent config. If we initialize on module init then we can't guarantee that the agent config has been loaded.

hush-hush commented on pkg/collector/python/datadog_agent.go:
```diff
@@ -203,3 +223,20 @@ func ObfuscateSQL(rawQuery *C.char, errResult **C.char) *C.char {
+func ObfuscateSQLExecPlan(jsonPlan *C.char, normalize C.bool, errResult **C.char) *C.char {
+	obfuscatedJSONPlan, err := lazyInitObfuscator().ObfuscateSQLExecPlan(
+		C.GoString(jsonPlan),
+		normalize == C.bool(true),
```
`bool(normalize)` doesn't work ?

djova replied:
um... yes, looks like it does! Didn't think of trying that. Updated.

hush-hush commented on rtloader/test/datadog_agent/datadog_agent.go:
```diff
@@ -247,3 +249,24 @@ func obfuscateSQL(rawQuery *C.char, errResult **C.char) *C.char {
+	shouldNormalize := normalize == C.bool(true)
+	switch C.GoString(rawQuery) {
+	case "raw-json-plan":
+		if shouldNormalize {
```
same here, why not `if bool(normalize) {` ?

djova replied:
Good point, updated

hush-hush reviewed (changes requested):
a few small nit but looks good otherwise !

djova commented on pkg/collector/python/datadog_agent.go:
```diff
@@ -187,14 +188,33 @@ func ReadPersistentCache(key *C.char) *C.char {
+// lazyInitObfuscator initializes the obfuscator the first time it is used. We can't initialize during the package init
+// because the obfuscator depends on config.Datadog and it isn't guaranteed to be initialized during package init, but
+// will definitely be initialized by the time one of the python checks runs
+func lazyInitObfuscator() *obfuscate.Obfuscator {
```
It is already using `sync.Once` to initialize the obfuscator inside this function. Is there a different way we should be using it?

hush-hush replied:
indeed, then we should not need the if statement.

djova replied:
Oh good point, removed!

djova commented on pkg/trace/obfuscate/obfuscate.go:
```diff
@@ -111,3 +139,59 @@ func compactWhitespaces(t string) string {
+		"used_key_parts",
+		"used_columns",
+		"ref",
+		"update",
```
> nit: can you order these alphabetically for each dbms?

Yes, ordered

> I'm having trouble locating an exhaustive list but I think these should also be included, a few I found going through mysql tests:

Added! Good finds. I didn't do an exhaustive search for the initial list.

djova commented on pkg/trace/obfuscate/obfuscate.go:
```diff
@@ -111,3 +139,59 @@ func compactWhitespaces(t string) string {
+	Enabled:         true,
+	TransformerType: "obfuscate_sql",
+	KeepValues: append([]string{
+		// mysql
```
Good catch, added.

djova commented on pkg/trace/config/apply.go:
```diff
@@ -31,6 +31,13 @@ type ObfuscationConfig struct {
 	// Mongo holds the obfuscation configuration for MongoDB queries.
 	Mongo JSONObfuscationConfig `mapstructure:"mongodb"`
 
+	// Mongo holds the obfuscation configuration for SQL Exec Plans. This is strictly for safety related obfuscation,
```
Good catch, fixed

knusbaum commented:
@djova @hush-hush Please have a look.
The code freeze for 7.21.0 is coming up tomorrow. If we'd like to get this in, we should merge. (not sure if macos_build is an issue)

djova commented:
> The code freeze for 7.21.0 is coming up tomorrow. If we'd like to get this in, we should merge. (not sure if macos_build is an issue)

@knusbaum We don't need to get this into the next release. There are still some minor changes that may get made so we're going to keep it open for a bit longer. 

Thanks for checking!

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -58,9 +79,21 @@ func (p *jsonObfuscator) setKey() {
 	p.wiped = false
 }
 
+func trimSideQuotes(quotedString []byte) []byte {
```
Would `strconv.Unquote` work here?

djova replied:
Oh didn't know about that. It works. Updated.

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -58,9 +79,21 @@ func (p *jsonObfuscator) setKey() {
 	var out strings.Builder
-	buf := make([]byte, 0, 10) // recording key token
+
+	keyBuf := make([]byte, 0, 10) // recording key token
```
We're going to need some numbers with `benchcmp` to see how this compares to the previous implementation. We know from profiles that on heavy sql spans this does weigh down on performance.

djova replied:
Here it is: 

```
benchcmp old.txt new.txt
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
ignoring BenchmarkObfuscateJSON/5003-8: before has 1 instances, after has 0
ignoring BenchmarkObfuscateJSON/437-8: before has 1 instances, after has 0
ignoring BenchmarkObfuscateJSON/671-8: before has 1 instances, after has 0
```

And the benchmarks themselves:
```
cat old.txt
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/obfuscate
BenchmarkObfuscateJSON/5003-8         	   24805	     47720 ns/op	   22120 B/op	     203 allocs/op
BenchmarkObfuscateJSON/437-8          	  191492	      6396 ns/op	    2952 B/op	      48 allocs/op
BenchmarkObfuscateJSON/671-8          	  175016	      6952 ns/op	    3016 B/op	      42 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/trace/obfuscate	4.387s

cat new.txt
goos: darwin
goarch: amd64
pkg: github.com/DataDog/datadog-agent/pkg/trace/obfuscate
BenchmarkObfuscateJSON/806-8         	   54447	     22840 ns/op	   10174 B/op	      80 allocs/op
BenchmarkObfuscateJSON/566-8         	   69775	     17706 ns/op	    8337 B/op	      67 allocs/op
BenchmarkObfuscateJSON/31-8          	  125083	     10101 ns/op	    5865 B/op	      35 allocs/op
PASS
ok  	github.com/DataDog/datadog-agent/pkg/trace/obfuscate	4.428s
```

This shouldn't have any impact on regular APM json obfuscation because the new value transformer code I added will only run when a value transformer is configured, which is only happening in the sql obfuscation code which is called from python checks. 

It can be configured to be used for APM though, so if we do want to obfuscate sql within json then we'll be able to do so.

djova replied:
Well it looks like everything is worse.

djova replied:
Looks like it's all in the initialization. If we move the obfuscator initialization outside of the benchmark then the benchmarks aren't so different:

```
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                          old ns/op     new ns/op     delta
BenchmarkObfuscateJSON/elasticsearch.body-8        48500         53455         +10.22%
BenchmarkObfuscateJSON/elasticsearch.body#01-8     6049          6368          +5.27%
BenchmarkObfuscateJSON/elasticsearch.body#02-8     6514          7253          +11.34%
BenchmarkObfuscateJSON/elasticsearch.body#03-8     6486          7418          +14.37%
BenchmarkObfuscateJSON/elasticsearch.body#04-8     1844          2066          +12.04%
BenchmarkObfuscateJSON/elasticsearch.body#05-8     1804          2121          +17.57%
BenchmarkObfuscateJSON/elasticsearch.body#06-8     1807          2112          +16.88%
BenchmarkObfuscateJSON/elasticsearch.body#07-8     1831          2052          +12.07%
BenchmarkObfuscateJSON/elasticsearch.body#08-8     748           802           +7.22%
BenchmarkObfuscateJSON/elasticsearch.body#09-8     918           1268          +38.13%
BenchmarkObfuscateJSON/elasticsearch.body#10-8     975           1084          +11.18%
BenchmarkObfuscateJSON/elasticsearch.body#11-8     1514          1896          +25.23%
BenchmarkObfuscateJSON/elasticsearch.body#12-8     1099          1250          +13.74%
BenchmarkObfuscateJSON/elasticsearch.body#13-8     1491          1685          +13.01%

benchmark                                          old allocs     new allocs     delta
BenchmarkObfuscateJSON/elasticsearch.body-8        195            195            +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#01-8     39             39             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#02-8     33             33             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#03-8     33             33             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#04-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#05-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#06-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#07-8     16             16             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#08-8     9              9              +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#09-8     9              9              +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#10-8     9              9              +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#11-8     14             14             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#12-8     11             11             +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#13-8     14             14             +0.00%

benchmark                                          old bytes     new bytes     delta
BenchmarkObfuscateJSON/elasticsearch.body-8        21702         21698         -0.02%
BenchmarkObfuscateJSON/elasticsearch.body#01-8     2456          2456          +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#02-8     2520          2520          +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#03-8     2520          2520          +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#04-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#05-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#06-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#07-8     712           712           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#08-8     296           296           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#09-8     328           328           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#10-8     328           328           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#11-8     616           616           +0.00%
BenchmarkObfuscateJSON/elasticsearch.body#12-8     504           488           -3.17%
BenchmarkObfuscateJSON/elasticsearch.body#13-8     616           616           +0.00%
```

djova replied:
Fixed here https://github.com/DataDog/datadog-agent/pull/5692/commits/0d73bfc16cd4184963480ca84c7e763d5dfc3e5b

gbbr replied:
Cool, I think it looks good. Thanks

gbbr commented on pkg/config/config.go:
```diff
@@ -697,8 +697,20 @@ func InitConfig(config Config) {
 	config.SetKnown("apm_config.obfuscation.mongodb.keep_values")
+	config.SetKnown("apm_config.obfuscation.mongodb.transform_values")
+	config.SetKnown("apm_config.obfuscation.mongodb.transformer_type")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan.enabled")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan.keep_values")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan.transform_values")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan.transformer_type")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan_normalize.enabled")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan_normalize.keep_values")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan_normalize.transform_values")
+	config.SetKnown("apm_config.obfuscation.sql_exec_plan_normalize.transformer_type")
```
Wouldn't it be better to group these under `apm_config.obfuscation.sql.*`? I see some other PRs attempting to add that and we may want more options later on. One possibility is `apm_config.obfuscation.sql.exec_plan.*`

djova replied:
I made it a bit simpler in https://github.com/DataDog/datadog-agent/commit/c5014db632ce98ccd97685e861f10007c58222dd 

We can't easily group them because each of these `(es, mongo, sql_exec_plan, and sql_exec_plan_normalize)` are configuring an independent json obfuscator. They can each have different keep_values and `obfuscate_sql_values`.

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -25,27 +27,47 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
 	for _, v := range cfg.KeepValues {
 		keepValue[v] = true
 	}
+	var transformValues map[string]bool
+	var transformer valueTransformer
```
I'd propose
```suggestion
	var (
	    transformVals map[string]bool
	    transformFunc valueTransformer
	)
```

djova replied:
Sure. It takes double the number of lines though 😛

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -25,27 +27,47 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
+	if cfg.TransformerType == "obfuscate_sql" {
+		transformer = o.ObfuscateSQLStringSafe
+	} else {
+		log.Warnf("unknown JSON value transformer type %s", cfg.TransformerType)
```
Can we cleanly check that all values are correct and warn about them in `apply.go`? If that's complicated, it's fine to leave it here, but ideally we'd validate everything there.

djova replied:
Simplified here https://github.com/DataDog/datadog-agent/commit/c5014db632ce98ccd97685e861f10007c58222dd

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -25,27 +27,47 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
+		log.Warnf("unknown JSON value transformer type %s", cfg.TransformerType)
+	}
+	if transformer != nil {
+		transformValues = make(map[string]bool, len(cfg.TransformValues))
```
Wouldn't it be easier for `cfg.TransformValues` to be a map from the git-go?

djova replied:
We could only if we make it a map in yaml, which I don't think is a good idea because this kind of setting seems nicer to have as a list in yaml. 

Better the user see 
`keys: ['a', 'b', 'c']`

than 

```
keys:
    a: true
    b: true
    c: true
```

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -25,27 +27,47 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
 	// as much of it as we could. It is safe to accept the output, even if partial.
 }
 
+type valueTransformer func(string) string
```
Is there any value on having this type? It's pretty straightforward. You can keep it if you like, but I'm thinking just naming things `var transformerFunc func(string) string` will be equally understandable, easy to type and read.

djova replied:
Good point, removed

gbbr commented on pkg/trace/obfuscate/obfuscate.go:
```diff
@@ -42,21 +46,45 @@ func (o *Obfuscator) SQLLiteralEscapes() bool {
+// LoadSQLObfuscator initializes the obfsucator from the standard agent config, setting defaults for SQLExecPlan
+// and SQLExecPlanNormalize if they are not set. Since this relies on config.Datadog it must be called after
+// config.Datadog has been initialized in order for config.Datadog to be used
+func LoadSQLObfuscator() *Obfuscator {
```
Why not `New*`?

djova replied:
That works, updated

gbbr commented on pkg/trace/obfuscate/sql.go:
```diff
@@ -342,3 +342,23 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
+func (o *Obfuscator) ObfuscateSQLStringSafe(sqlString string) string {
+	result, err := o.ObfuscateSQLString(sqlString)
+	if err != nil {
+		log.Debugf("failed to obfuscate sql string '%s': %s", sqlString, err.Error())
```
```suggestion
		log.Debugf("Failed to obfuscate SQL string '%s': %s", sqlString, err.Error())
```
We had a PR a while back which made all log entries capitalised.

djova replied:
Updated

gbbr commented on pkg/trace/obfuscate/sql.go:
```diff
@@ -342,3 +342,23 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
+	result, err := o.ObfuscateSQLString(sqlString)
+	if err != nil {
+		log.Debugf("failed to obfuscate sql string '%s': %s", sqlString, err.Error())
+		return "datadog-agent failed to obfuscate sql string. enable agent debug logs for more info."
```
Is there any point in putting `err.Error()` here?

djova replied:
No, the point is that this can only return a string. If there's an error then we put the error message in the string itself. That way when viewing the query that failed to obfuscate in our product it's clear that there's a problem and it's not just empty or missing. 
https://github.com/DataDog/datadog-agent/pull/5692/commits/65e0e51eff80d093aa80995bdb7c87e43007b628

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -26,26 +27,42 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
 
 type jsonObfuscator struct {
-	keepers map[string]bool // these keys will not be obfuscated
+	keepKeys      map[string]bool // these keys will not be obfuscated
+	transformKeys map[string]bool // these keys pass through the transformer
```
```suggestion
	keepKeys      map[string]bool // the values for these keys will not be obfuscated
	transformKeys map[string]bool // the values for these keys pass through the transformer
```
I guess this should help a bit?

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -26,26 +27,42 @@ func (o *Obfuscator) obfuscateJSON(span *pb.Span, tag string, obfuscator *jsonOb
 		keepValue[v] = true
 	}
+	var (
+		transformValues map[string]bool
```
```suggestion
		transformKeys map[string]bool
```
Otherwise we might get confused a bit, no?

gbbr commented on pkg/trace/obfuscate/json.go:
```diff
@@ -87,15 +108,29 @@ func (p *jsonObfuscator) obfuscate(data []byte) (string, error) {
+					v = string(valBuf)
+				}
+				result := p.transformer(v)
+				out.Write([]byte(`"`))
```
Can we use `WriteByte` and `WriteString` here?

djova replied:
Ah didn't know those existed. Updated

gbbr reviewed (commented):
This looks good but I've only now realised that this is enabled by default. That means we've now made the obfuscator ~10% slower.

djova commented:
> This looks good but I've only now realised that this is enabled by default. That means we've now made the obfuscator ~10% slower.

I reran the benchmarks we have here with all of the new code commented out and there isn't a clear 10% increase. It seems that there's a +/- 20% even across runs on the same code path. The new sql obfuscation doesn't run at all unless you're enabling sql obfuscation. So the only thing we're adding here is a couple of conditional nill/bool checks per key, which should be negligible. 

```
BenchmarkObfuscateJSON/elasticsearch.body-8                                   124116        187878        +51.37%
BenchmarkObfuscateJSON/elasticsearch.body#01-8                                13924         31640         +127.23%
BenchmarkObfuscateJSON/elasticsearch.body#02-8                                15547         21335         +37.23%
BenchmarkObfuscateJSON/elasticsearch.body#03-8                                16696         22763         +36.34%
BenchmarkObfuscateJSON/elasticsearch.body#04-8                                4316          4124          -4.45%
BenchmarkObfuscateJSON/elasticsearch.body#05-8                                3669          4953          +35.00%
BenchmarkObfuscateJSON/elasticsearch.body#06-8                                4607          3432          -25.50%
BenchmarkObfuscateJSON/elasticsearch.body#07-8                                3156          3351          +6.18%
BenchmarkObfuscateJSON/elasticsearch.body#08-8                                1265          1279          +1.11%
BenchmarkObfuscateJSON/elasticsearch.body#09-8                                2237          1557          -30.40%
BenchmarkObfuscateJSON/elasticsearch.body#10-8                                1852          1756          -5.18%
BenchmarkObfuscateJSON/elasticsearch.body#11-8                                2943          2994          +1.73%
BenchmarkObfuscateJSON/elasticsearch.body#12-8                                1919          1937          +0.94%
BenchmarkObfuscateJSON/elasticsearch.body#13-8                                2697          3217          +19.28%
```

I also collected CPU profiles for the benchmark before & after and didn't see any obvious change. Most of the time is spent manipulating strings.

gbbr commented on pkg/trace/obfuscate/obfuscate.go:
```diff
@@ -42,21 +46,45 @@ func (o *Obfuscator) SQLLiteralEscapes() bool {
 	return &o
 }
 
+// NewSQLObfuscator initializes the obfsucator from the standard agent config, setting defaults for SQLExecPlan
```
Since this is only used in `pkg/collector/python`, should we move this function over there along with the defaults?

djova replied:
I don't think so, because most of the stuff in `pkg/collector/python` is cgo wrapper code calling out to core logic is that elsewhere. So I think it fits pretty well in here.

gbbr replied:
I don't like this because of two reasons:
1. It confuses the API by having the name `NewSQLObfuscator`, which can lead one to believe it's an SQL obfuscator, which it's not: it's in fact a span obfuscator which supports everything else (Redis, ES, etc) but has special behaviour for JSON obfuscators.
2. It's not used anywhere inside `pkg/trace`
3. Seems to belong in `pkg/collector/python`

gbbr replied:
I still hold the above belief and am not sure I understand why this is here...

djova replied:
On second thought, I agree. Moved it https://github.com/DataDog/datadog-agent/pull/5692/commits/ba27f472e93cdcbc8dbe0f9cd9eff28119dff4dd

gbbr commented on pkg/config/config.go:
```diff
@@ -697,8 +697,17 @@ func InitConfig(config Config) {
 	config.SetKnown("apm_config.obfuscation.elasticsearch.enabled")
 	config.SetKnown("apm_config.obfuscation.elasticsearch.keep_values")
+	config.SetKnown("apm_config.obfuscation.elasticsearch.transform_values")
+	config.SetKnown("apm_config.obfuscation.elasticsearch.transformer_type")
```
I think you forgot this here by accident?

djova replied:
Yes, removed.

gbbr commented:
I guess this has some conflicts now. @djova we're freezing Friday. Does this need to go in? If not, I'll change the milestone.

djova commented:
> I guess this has some conflicts now. @djova we're freezing Friday. Does this need to go in? If not, I'll change the milestone.

It doesn't need to go in for this milestone. Let's push back to the next one. Thanks!

gbbr commented on pkg/trace/obfuscate/sql.go:
```diff
@@ -358,3 +358,25 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
+
+// ObfuscateSQLStringSafe obfuscates the sql string, returning either the successfully obfuscated string or a string
+// containing an error message explaining why the obfuscation failed
+func (o *Obfuscator) ObfuscateSQLStringSafe(sqlString string) string {
```
There are way too many of these. We have:
* `obfuscateSQL`
* `obfuscateSQLString`
* `ObfuscateSQLString`

I think that's enough :) 

Instead, let's add a method onto `jsonObfuscator` called perhaps `obfuscateSQL`:

```go
// sqlObfuscator returns an SQL obfuscation function which uses o as its Obfuscator. It returns
// either the successfully obfuscated string or a string containing an error message explaining
// why the obfuscation failed.
func (*jsonObfuscator) sqlObfuscator(o *Obfuscator) func(string) string {
    return func(in string) string {
        // ... contents of this function
    }
}
```
And create this as the transformer.

What do you think?

djova replied:
I agree that there are many of them but I'm not sure that this approach would make things clearer. If anything it may get more convoluted because it's a function called `sqlObfuscator` that takes an `obufscator` and presumably returns an "obfuscator" that's not actually an obfuscator in the way the word is used in the rest of this module, rather it's a function that obfuscates only strings. 🤷‍♂️

gbbr replied:
I understand, but this is way too much, and it makes no sense to have one more. This method belongs on `*jsonObfuscator`. If you can find an alternative way that's fine but I feel pretty strongly about moving it away from here. I don't find it that confusing. Call it `tranformerFunc` if you will :)

djova replied:
Ok, I see what you mean. Refactored here https://github.com/DataDog/datadog-agent/pull/5692/commits/d7064f556056480bfe5d020295f2e5a9dea2f2e9 does that work?

djova commented on pkg/trace/obfuscate/obfuscate.go:
```diff
@@ -11,16 +11,20 @@ import (
 
-	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	traceconfig "github.com/DataDog/datadog-agent/pkg/trace/config"
```
Renamed

djova commented on releasenotes/notes/sql-exec-plan-obfuscation-916c3d66e436d7ce.yaml:
```diff
@@ -0,0 +1,15 @@
+---
+features:
+  - |
+    New `datadog_agent.obfuscate_sql_exec_plan` function exposed to python
```
Added

djova commented on releasenotes/notes/sql-exec-plan-obfuscation-916c3d66e436d7ce.yaml:
```diff
@@ -0,0 +1,15 @@
+    New `datadog_agent.obfuscate_sql_exec_plan` function exposed to python
+    checks to enable obfuscation of json-encoded SQL Query Execution Plans.
+  - |
+    New `jsonObfuscator` feature enabling obfuscation of SQL queries contained
```
I updated the note to 

```
    APM: new `obfuscate_sql_values` option in `apm_config.obfuscation` enabling optional obfuscation
    of SQL queries contained in JSON data collected from some APM services (ES & Mongo)
```

It technically is public facing now. So anyone that is doing json obfuscation on Mongo, for example, will be able to do enable SQL obfuscation for some values if they want.

gbbr commented on pkg/collector/python/datadog_agent.go:
```diff
@@ -8,6 +8,8 @@
 package python
 
 import (
+	traceconfig "github.com/DataDog/datadog-agent/pkg/trace/config"
```
This may be in the wrong imports group

djova replied:
Reordered
